### PR TITLE
perf(lazy-pages,runtime-interface): optimise memory access and usage

### PR DESCRIPTION
**Pull request by ukint-vs**
Resolves # .

-
-
-

@reviewer-or-team

### CODE REVIEW

**Review by breathx — APPROVED** [review]

**Comment by NikVolf** [comment]
rebase to master and request Validator testing?

**Comment by ukint-vs** [comment]
> rebase to master and request Validator testing?

it's not done. The weight of per-byte becnhes has been increased. Iterators need to be profiled or removed.

File: runtime-interface/src/lib.rs [hunk 1]
```diff
@@ -114,49 +105,50 @@ pub trait GearRI {
     ) -> (GasLeft, Result<(), ProcessAccessErrorVer1>) {
         let mut gas_left = gas_left.0;
         let gas_before = gas_left.gas;
-        let res = lazy_pages::pre_process_memory_accesses(reads, writes, &mut gas_left.gas);
+        let res = lazy_pages::pre_process_memory_accesses(
+            &mut reads.iter().cloned(),
```
[the remark this answers is not in the document]
Comment (ukint-vs) [reply runtime-interface/src/lib.rs:103]: it's deprecated api.

File: lazy-pages/src/host_func.rs [hunk 2]
```diff
@@ -96,41 +143,32 @@ impl<'a> AccessHandler for HostFuncAccessHandler<'a> {
...
+        },
+        page_iter,
+        page_size,
+    )
 }
 
 pub fn pre_process_memory_accesses(
-    reads: &[MemoryInterval],
-    writes: &[MemoryInterval],
+    reads: &mut (impl Iterator<Item = MemoryInterval> + std::fmt::Debug),
```
Comment (breathx) [inline lazy-pages/src/host_func.rs:170]: Why should `Debug` appear here?

File: lazy-pages/src/host_func.rs [hunk 3]
```diff
@@ -26,19 +26,69 @@ use crate::{
...
+    pages::{GearPage, PageDynSize, PageNumber},
 };
 use gear_lazy_pages_common::{ProcessAccessError, Status};
-use std::collections::BTreeSet;
+use std::ops::RangeInclusive;
+
+pub struct MemoryIntervalPageIterator<'a> {
+    intervals: &'a mut dyn Iterator<Item = MemoryInterval>,
+    page_size: u32,
+    current_pages: Option<RangeInclusive<u32>>,
```
Comment (breathx) [inline lazy-pages/src/host_func.rs:37]: why isn't it btreeset?

Comment (breathx) [reply lazy-pages/src/host_func.rs:37]: or anything else

File: lazy-pages/src/host_func.rs [hunk 4]
```diff
@@ -96,41 +143,32 @@ impl<'a> AccessHandler for HostFuncAccessHandler<'a> {
     }
 }
 
-fn accesses_pages(
+fn process_memory_intervals(
     ctx: &mut LazyPagesExecutionContext,
-    accesses: &[MemoryInterval],
-    pages: &mut BTreeSet<GearPage>,
-) -> Result<(), Error> {
+    intervals: &mut impl Iterator<Item = MemoryInterval>,
```
Comment (breathx) [inline lazy-pages/src/host_func.rs:148]: why should they be mutable?

**Review by breathx — CHANGES_REQUESTED** [review]

**Comment by breathx** [comment]
Please reopen once work will be continued

**Pull request closed without merging**